Sources/MlemMiddleware/Content Models/SubscriptionList/SubscriptionList.swift

```diff
@@ -0,0 +1,102 @@
+//
+//  SubscriptionList.swift
+//
+//
+//  Created by Sjmarf on 05/05/2024.
+//
+
+import SwiftUI
+
+@Observable
+public class SubscriptionList {
+    /// All subscribed-to communities, including favorited communities.
+    public private(set) var communities: Set<Community2> = .init()
+    public private(set) var favorites: [Community2] = .init()
+    public private(set) var alphabeticSections: [String?: [Community2]] = .init()
+    
+    var favoriteIDs: Set<Int> {
+        get { getFavorites() }
+        set { self.setFavorites(newValue) }
+    }
+    
+    private var getFavorites: () -> Set<Int>
+    private var setFavorites: (Set<Int>) -> Void
+    
+    private var apiClient: ApiClient
+    
+    internal init(
+        apiClient: ApiClient,
+        getFavorites: @escaping () -> Set<Int>,
+        setFavorites: @escaping (Set<Int>) -> Void) {
+        self.apiClient = apiClient
+        self.getFavorites = getFavorites
+        self.setFavorites = setFavorites
+    }
+    
+    public func refresh() async throws {
+        _ = try await apiClient.getSubscriptionList()
+    }
+    
+    private func categoryForCommunity(_ community: Community2) -> String? {
+        let first = String(community.name.first ?? "#").folding(options: .diacriticInsensitive, locale: .current)
+        guard first.first?.isLetter ?? false else { return nil }
+        return first.uppercased()
+    }
+    
+    @MainActor
+    internal func updateCommunities(with communities: Set<Community2>) {
+        self.communities = communities
+        let sections: [String?: [Community2]] = .init(
+            grouping: communities,
+            by: { categoryForCommunity($0) }
+        )
+        for section in sections {
+            self.alphabeticSections[section.key] = section.value.sorted(by: { $0.name < $1.name })
+        }
+    }
+    
+    func updateCommunitySubscription(community: Community2) {
+        if community.subscribed {
+            if !self.communities.contains(community) {
+                self.addCommunity(community: community)
+            }
+            if self.favoriteIDs.contains(community.id) != community.shouldBeFavorited {
+                if community.shouldBeFavorited {
+                    self.favoriteIDs.insert(community.id)
+                    let index = self.favorites.insertionIndex { $0.name < community.name }
+                    self.favorites.insert(community, at: index)
+                } else {
+                    self.favoriteIDs.remove(community.id)
+                    if let index = favorites.firstIndex(of: community) {
+                        favorites.remove(at: index)
+                    }
```

Comment:
Is there a reason not to just use `.removeFirst(community)`?

Reply:
Unfortunately there isn't a better way to do this, unless we define one in an extension. `removeFirst(n)`  removes the first `n` elements from the array.